[Swift in WebKit] Add Swift to the style checker

rr-codes opened the pull request:
#### 87915cf38e4831154edea8c6ddf9902e5baae27c
<pre>
[Swift in WebKit] Add Swift to the style checker
<a href="https://bugs.webkit.org/show_bug.cgi?id=295410">https://bugs.webkit.org/show_bug.cgi?id=295410</a>
<a href="https://rdar.apple.com/154964698">rdar://154964698</a>

Reviewed by Abrar Rahman Protyasha.

Use `swift format` to check for formatting and linting issues in Swift code.

* Tools/Scripts/webkitpy/style/checker.py:
(FileType):
(CheckerDispatcher._file_type):
(CheckerDispatcher._create_checker):
* Tools/Scripts/webkitpy/style/checkers/swift.py: Added.
(SwiftChecker):
(SwiftChecker.__init__):
(SwiftChecker._process_lines):
(SwiftChecker.check):
* Tools/Scripts/webkitpy/style/checkers/swift_unittest.py: Added.
(SwiftCheckerTest):
(SwiftCheckerTest.test_init):
(SwiftCheckerTest.test_init._mock_handle_style_error):
(SwiftCheckerTest.test_check):
(SwiftCheckerTest.test_check._mock_handle_style_error):
* Tools/Scripts/webkitpy/style/checkers/swift_unittest_input.swift: Added.
(foo):

Canonical link: <a href="https://commits.webkit.org/297011@main">https://commits.webkit.org/297011@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/244e857e64f8e2b1ff6d7e3eaf4c4c417b27e7a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110232 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20323 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116256 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60480 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30569 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38478 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83807 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24387 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64254 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/109664 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23748 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17399 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60050 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93763 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119046 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37272 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92779 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37644 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92605 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23605 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37599 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33161 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37166 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36828 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38537 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,26 @@
+import re
```
We should add the copyright header to this new file

briannafan commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class SwiftChecker(object):
+    def __init__(self, file_path: str, handle_style_error):
```
nit: If we use typed Python, let's make it consistent throughout the file, or we can just remove `str` here.

briannafan commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,26 @@
+        for line in lint_result.stderr.splitlines():
+            match = re.match(pattern, line)
+
+            filename = match.group(1)
```
nit: Doesn't seem like we're actually using this variable?

aprotyas commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,26 @@
+        lint_result = subprocess.run([f'/usr/bin/swift', 'format', 'lint', '--strict', file_path], capture_output=True, text=True)
+
+        # matches <filename>:<line>: error: [<category>] <message>
+        pattern = r'^([^:]+):(\d+):\d+:\s*(warning|error):\s*\[([^\]]+)\]\s*(.+)$'
```
Probably worth compiling this since you're not reconfiguring the pattern at any point

aprotyas replied:
```py
pattern = re.compile(r'^([^:]+):(\d+):\d+:\s*(warning|error):\s*\[([^\]]+)\]\s*(.+)$')
for line in lint_result.stderr.splitlines():
    match = pattern.match(line)
```

briannafan reviewed (changes requested):
Looks good generally! I'd like to see some unit tests for `SwiftChecker,` similar to [style/checkers/python_unittest.py](https://github.com/WebKit/WebKit/blob/main/Tools/Scripts/webkitpy/style/checkers/python_unittest.py) or any of the other checker unit tests.

aprotyas commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,26 @@
+            filename = match.group(1)
+            line_number = match.group(2)
+            category = match.group(4)
+            message = match.group(5)
```
Consider using named groups? Helps readability of the regex expression anyway

```py
pattern = re.compile(
    r'^(?P<filename>[^:]+):'
    r'(?P<line_number>\d+):'
    r'\d+:'
    r'\s*(?P<category>warning|error):'
    r'\s*\[(?P<subcategory>[^\]]+)\]\s*'
    r'(?P<message>.+)$'
)
```

Then you can do

```
    filename = match.group("filename")
    line_number = match.group("line_number")
    category = match.group("category")
    subcategory = match.group("subcategory")
    message = match.group("message")
```

aprotyas commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,26 @@
+        # matches <filename>:<line>: error: [<category>] <message>
+        pattern = r'^([^:]+):(\d+):\d+:\s*(warning|error):\s*\[([^\]]+)\]\s*(.+)$'
+        for line in lint_result.stderr.splitlines():
+            match = re.match(pattern, line)
```
You could early exit per loop iteration if nothing matches

```py
if not match:
    continue
```

aprotyas commented on Tools/Scripts/webkitpy/style/checkers/swift_unittest.py:
```diff
@@ -0,0 +1,65 @@
+        checker = SwiftChecker(file_path, _mock_handle_style_error)
+        checker.check(lines=[])
+
+        # FIXME: https://bugs.webkit.org/show_bug.cgi?id=204133
```
What's this FIXME about? This file is Python 3 right 😛

aprotyas commented on Tools/Scripts/webkitpy/style/checkers/swift.py:
```diff
@@ -0,0 +1,59 @@
+        self.handle_style_error = handle_style_error
+
+    def _process_lines(self, file_path, lines, error):
+        lint_result = subprocess.run([f'/usr/bin/swift', 'format', 'lint', '--strict', file_path], capture_output=True, text=True)
```
f-string not needed here!

```suggestion
        lint_result = subprocess.run(['/usr/bin/swift', 'format', 'lint', '--strict', file_path], capture_output=True, text=True)
```

rr-codes committed:
[Swift in WebKit] Add Swift to the style checker
https://bugs.webkit.org/show_bug.cgi?id=295410
rdar://154964698

Reviewed by Abrar Rahman Protyasha.

Use `swift format` to check for formatting and linting issues in Swift code.

* Tools/Scripts/webkitpy/style/checker.py:
(FileType):
(CheckerDispatcher._file_type):
(CheckerDispatcher._create_checker):
* Tools/Scripts/webkitpy/style/checkers/swift.py: Added.
(SwiftChecker):
(SwiftChecker.__init__):
(SwiftChecker._process_lines):
(SwiftChecker.check):
* Tools/Scripts/webkitpy/style/checkers/swift_unittest.py: Added.
(SwiftCheckerTest):
(SwiftCheckerTest.test_init):
(SwiftCheckerTest.test_init._mock_handle_style_error):
(SwiftCheckerTest.test_check):
(SwiftCheckerTest.test_check._mock_handle_style_error):
* Tools/Scripts/webkitpy/style/checkers/swift_unittest_input.swift: Added.
(foo):

Canonical link: https://commits.webkit.org/297011@main
